server/src/main/java/org/elasticsearch/health/node/DiskHealthIndicatorService.java

```diff
@@ -0,0 +1,284 @@
+            Set<DiscoveryNodeRole> roles = clusterService.state()
+                .nodes()
+                .getNodes()
+                .values()
+                .stream()
+                .filter(node -> problemNodes.contains(node.getId()))
+                .map(DiscoveryNode::getRoles)
+                .flatMap(Collection::stream)
+                .collect(Collectors.toSet());
+            if (roles.contains(DiscoveryNodeRole.MASTER_ROLE)) {
+                impacts.addAll(StableMasterHealthIndicatorService.UNSTABLE_MASTER_IMPACTS);
+            }
+            if (roles.contains(DiscoveryNodeRole.INGEST_ROLE)) {
+                impacts.add(HEALTH_INGEST_NODES_PIPELINES_IMPACT);
+                impacts.add(HEALTH_INGEST_NODES_MONITORING_IMPACT);
+            }
+            if (roles.contains(DiscoveryNodeRole.ML_ROLE)) {
+                impacts.add(HEALTH_ML_NODES_IMPACT);
+            }
+            if (roles.contains(DiscoveryNodeRole.REMOTE_CLUSTER_CLIENT_ROLE)) {
+                impacts.add(HEALTH_REMOTE_CLUSTER_CLIENT_NODES_IMPACT);
+            }
+            if (roles.contains(DiscoveryNodeRole.TRANSFORM_ROLE)) {
+                impacts.add(HEALTH_TRANSFORM_NODES_IMPACT);
+            }
+            diagnosisList = getDiskProblemsDiagnosis(explain, problemNodes, problemIndices);
+        }
+        return createIndicator(healthStatus, symptom, details, impacts, diagnosisList);
+    }
+
+    private Set<String> getIndicesForNodes(Set<String> nodes) {
+        return clusterService.state()
+            .routingTable()
+            .allShards()
+            .stream()
+            .filter(routing -> nodes.contains(routing.currentNodeId()))
+            .map(routing -> routing.index().getName())
+            .collect(Collectors.toSet());
+    }
+
+    /**
+     * This method logs if any nodes in the cluster state do not have health info results reported. This is logged at debug level and is
+     * not ordinarly important, but could be useful in tracking down problems where nodes have stopped reporting health node information.
+     * @param diskHealthInfoMap A map of nodeId to DiskHealthInfo
+     */
+    private void logMissingHealthInfoData(Map<String, DiskHealthInfo> diskHealthInfoMap) {
+        if (logger.isDebugEnabled()) {
+            Set<String> nodeIdsInClusterState = clusterService.state()
+                .nodes()
+                .stream()
+                .map(DiscoveryNode::getId)
+                .collect(Collectors.toSet());
+            Set<String> nodeIdsInHealthInfo = diskHealthInfoMap.keySet();
+            if (nodeIdsInHealthInfo.containsAll(nodeIdsInClusterState) == false) {
+                String problemNodes = nodeIdsInClusterState.stream()
+                    .filter(node -> nodeIdsInHealthInfo.contains(node) == false)
+                    .collect(Collectors.joining(", "));
+                logger.debug("The following nodes are in the cluster state but not reporting health data: [{}}]", problemNodes);
```

Comment:
To make it easier to debug I would suggest to use both name and id in the log line. I think it's worth to use a `for` loop to facilitate that. What do you think?

Reply:
Good idea.